[Excel, Word] (Bindings) Refresh the Bindings Common API Article (#5306)
OfficeDev/office-js-docs-pr: docs/develop/bind-to-regions-in-a-document-or-spreadsheet.md

[!include[information about the common API](../includes/alert-common-api-info.md)]

Bindings let your add-in consistently access specific regions of a document or spreadsheet. Think of a binding as a bookmark that remembers a specific location, even if users change their selection or navigate elsewhere in the document. Specifically, here are what bindings offer your add-in.

- **Access common data structures** across supported Office applications, such as tables, ranges, or text.
- **Read and write data** without requiring users to make a selection first.
- **Create persistent relationships** between your add-in and document data. Bindings are saved with the document and work across sessions.

To create a binding, call one of these methods to associate a document region with a unique identifier: [addFromPromptAsync], [addFromSelectionAsync], or [addFromNamedItemAsync]. Once you've established the binding, use its identifier to read from or write to that region anytime.

You can also subscribe to data and selection change events for specific bound regions. This means your add-in only gets notified about changes within the bound area, not the entire document.

> ## Work with bindings

Reviewer: This heading seems weak. For SEO, considering eliminating as the primary H2 in this topic, rewriting, or combining with "Choose the right binding type" heading.

}
```

The binding type is text, so a [TextBinding] will be created for the user's selection in the prompt.

The second parameter contains the ID of the new binding. If you don't specify an ID, one is generated automatically.

The anonymous function passed as the third _callback_ parameter runs when the binding creation is complete. When the callback function runs, the [AsyncResult] object contains the call's status and the newly created binding.

> The following figure shows the built-in range selection prompt in Excel.

Reviewer: ```suggestion
The following screenshot shows the built-in range selection prompt in Excel.
```

> ![The Select Data dialog.](../images/agave-api-overview-excel-selection-ui.png)

Reviewer: I noticed lots of images are named agave- something. This is minor but it is a wasted SEO opportunity. If possible, I recommend renaming the files to office-add-in something.

```

**For Excel**, the `itemName` parameter of [addFromNamedItemAsync] can refer to an existing named range, a range specified with A1 reference style (`"A1:A3"`), or a table. By default, Excel assigns the names "Table1" for the first table, "Table2" for the second table, and so on. To assign a meaningful name to a table in the Excel UI, use the **Table Name** property on the **Table Tools | Design** tab.

> [!NOTE]
> In Excel, when specifying a table as a named item, you must fully qualify the name to include the worksheet name in this format (e.g., `"Sheet1!Table1"`).

> This example creates a binding in Excel to the first three cells in column A (`"A1:A3"`), assigns the ID `"MyCities"`, and then writes three city names to that binding.

Reviewer: ```suggestion
The following function creates a binding in Excel to the first three cells in column A (`"A1:A3"`), assigns the ID `"MyCities"`, and then writes three city names to that binding.
```

}
```

The anonymous function passed as the `callback` parameter runs when the operation is complete. The function is called with a single parameter, `asyncResult`, which contains an array of the bindings in the document. The array is iterated to build a string that contains the IDs of the bindings. The string is then displayed in a message box.

## Get a binding by ID using getByIdAsync

> This example shows how to use the [getByIdAsync] method to get a binding in a document by specifying its ID. This example assumes that a binding named `'myBinding'` was added to the document using one of the methods described earlier in this article.

Reviewer: ```suggestion
The following example uses the [getByIdAsync] method to get a binding in a document by specifying its ID. This example assumes that a binding named `'myBinding'` was added to the document using one of the methods described earlier in this article.
```

> In the example, the first `id` parameter is the ID of the binding to retrieve.

Reviewer: ```suggestion
In this example, the first `id` parameter is the ID of the binding to retrieve.
```

The anonymous function passed as the second _callback_ parameter runs when the operation is completed. The function is called with a single parameter, _asyncResult_, which contains the call's status and the binding with the ID "myBinding".

## Get a binding by ID using `Office.select`

> This example shows how to use the [Office.select] function to get a [Binding] object promise in a document by specifying its ID in a selector string. It then calls the [getDataAsync] method to get data from the specified binding. This example assumes that a binding named `'myBinding'` was added to the document using one of the methods described earlier in this article.

Reviewer: ```suggestion
The following example uses the [Office.select] function to get a [Binding] object promise in a document by specifying its ID in a selector string. It then calls the [getDataAsync] method to get data from the specified binding. This example assumes that a binding named `'myBinding'` was added to the document using one of the methods described earlier in this article.
```

```

> [!NOTE]

> > If the `select` function promise successfully returns a [Binding] object, that object exposes only the following four methods: [getDataAsync], [setDataAsync], [addHandlerAsync], and [removeHandlerAsync]. If the promise can't return a Binding object, the `onError` callback can be used to access an [asyncResult].error object to get more information. If you need to call a member of the Binding object other than the four methods exposed by the [Binding] object promise returned by the `select` function, instead use the [getByIdAsync] method by using the [Document.bindings] property and [getByIdAsync] method to retrieve the [Binding] object.

Reviewer: This doesn't need to be a note. Consider changing to a normal paragraph describing the previous example.

## Release a binding by ID

> This example shows how to use the [releaseByIdAsync] method to release a binding in a document by specifying its ID.

Reviewer: ```suggestion
The following example uses the [releaseByIdAsync] method to release a binding in a document by specifying its ID.
```

> In the example, the first `id` parameter is the ID of the binding to release.

Reviewer: ```suggestion
In this example, the first `id` parameter is the ID of the binding to release.
```

}
```

`myBinding` is a variable that contains an existing text binding in the document. Alternatively, you could use [Office.select] to access the binding by its ID, and start your call to the [getDataAsync] method, like this:

```js
Office.select("bindings#myBindingID").getDataAsync
```

The anonymous function passed into the method is a callback that runs when the operation is complete. The [AsyncResult].value property contains the data within `myBinding`. The type of the value depends on the binding type. The binding in this example is a text binding, so the value will contain a string. For additional examples of working with matrix and table bindings, see the [getDataAsync] method topic.

## Write data to a binding

> This example shows how to use the [setDataAsync] method to set data in an existing binding.

Reviewer: ```suggestion
The following example uses the [setDataAsync] method to set data in an existing binding.
```

> In the example, the first parameter is the value to set on `myBinding`. Because this is a text binding, the value is a `string`. Different binding types accept different types of data.

Reviewer: ```suggestion
In this example, the first parameter is the value to set on `myBinding`. Because this is a text binding, the value is a `string`. Different binding types accept different types of data.
```

The anonymous function passed into the method is a callback that runs when the operation is complete. The function is called with a single parameter, `asyncResult`, which contains the result's status.

## Detect changes to data or selection in a binding

> This example shows how to attach an event handler to the [DataChanged](/javascript/api/office/office.binding) event of a binding with an ID of "MyBinding".

Reviewer: ```suggestion
The following function attaches an event handler to the [DataChanged](/javascript/api/office/office.binding) event of a binding with an ID of "MyBinding".
```

The `myBinding` is a variable that contains an existing text binding in the document.

The first _eventType_ parameter of [addHandlerAsync] specifies the name of the event to subscribe to. [Office.EventType] is an enumeration of available event type values. `Office.EventType.BindingDataChanged` evaluates to the string "bindingDataChanged".

The `dataChanged` function passed as the second _handler_ parameter is an event handler that runs when the data in the binding is changed. The function is called with a single parameter, _eventArgs_, which contains a reference to the binding. This binding can be used to retrieve the updated data.

Similarly, you can detect when a user changes selection in a binding by attaching an event handler to the [SelectionChanged] event of a binding. To do that, specify the `eventType` parameter of [addHandlerAsync] as `Office.EventType.BindingSelectionChanged` or `"bindingSelectionChanged"`.

> You can add multiple event handlers for a given event by calling [addHandlerAsync] again and passing in an additional event handler function for the `handler` parameter. This will work correctly as long as the name of each event handler function is unique.

Reviewer: ```suggestion
You can add multiple event handlers for a given event by calling [addHandlerAsync] again and passing in an additional event handler function for the `handler` parameter. This works correctly as long as the name of each event handler function is unique.
```
otherwise, it doesn't work correctly?

### Remove an event handler

> To remove an event handler for an event, call [removeHandlerAsync] passing in the event type as the first _eventType_ parameter, and the name of the event handler function to remove as the second _handler_ parameter. For example, the following function will remove the `dataChanged` event handler function added in the previous section's example.

Reviewer: ```suggestion
To remove an event handler for an event, call [removeHandlerAsync] passing in the event type as the first _eventType_ parameter, and the name of the event handler function to remove as the second _handler_ parameter. For example, the following function removes the `dataChanged` event handler function added in the previous section's example.
```
